pkg/middleware/writer.go

```diff
@@ -67,6 +67,13 @@ func (r *CustomWriter) WriteHeader(code int) {
 	r.statusCode = code
 }
 
+// SetHeader will set the response HTTP header key/value pair
+func (r *CustomWriter) SetHeader(key, value string) {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+	r.Headers.Set(key, value)
+}
+
```

Comment:
don't see the goal of that, the `Header()` function already make a proxy to the base response writer Header if the response has not been send, write returns an empty Header instance otherwise. And already call the mutex.

Reply:
The `http.Header` is a map, which is a reference type. It's being accessed from multiple goroutines. I still think there's something wrong elsewhere because we're still passing the `r.Rw.Headers()`, but I can't chase that today.